File: go-test/internal/api/handlers/analytics.go
package handlers

import (
	"sort"

	"github.com/gin-gonic/gin"
)

func AvailableCategoriesHandler(c *gin.Context) {
	mu.RLock()
	defer mu.RUnlock()

	categoryMap := make(map[string]bool)
	for _, p := range allProducts {
		if p.Category != "" {
			categoryMap[p.Category] = true
		}
	}

	categories := make([]string, 0, len(categoryMap))
	for cat := range categoryMap {
		categories = append(categories, cat)
	}

	sort.Strings(categories)

	c.JSON(200, gin.H{
		"categories": categories,
		"total":      len(categories),
	})
}

func CategoriesStatsHandler(c *gin.Context) {
	mu.RLock()
	defer mu.RUnlock()

	stats := calculateCategoryStats()
	c.JSON(200, stats)
}

func CategoryDetailHandler(c *gin.Context) {
	categoryName := c.Param("name")

	mu.RLock()
	defer mu.RUnlock()

	products := filterByCategory(allProducts, categoryName)

	c.JSON(200, gin.H{
		"category": categoryName,
		"products": products,
		"stats": gin.H{
			"totalProducts":   len(products),
			"stores":          getStoreBreakdown(products),
			"priceComparison": getPriceComparison(products),
		},
	})
}

func CheapestProductsHandler(c *gin.Context) {
	mu.RLock()
	defer mu.RUnlock()

	result := make(map[string][]gin.H)
	stats := calculateCategoryStats()

	for category, catStats := range stats {
		var cheapest []gin.H
		products := filterByCategory(allProducts, category)

		for _, p := range products {
			if p.Store == catStats.CheapestStore {
				cheapest = append(cheapest, gin.H{
					"name":     p.Name,
					"price":    p.Price,
					"store":    p.Store,
					"link":     p.Link,
					"image":    p.Image,
					"category": p.Category,
				})
			}
		}

		result[category] = cheapest
	}

	c.JSON(200, result)
}

func StoresComparisonHandler(c *gin.Context) {
	mu.RLock()
	defer mu.RUnlock()

	supermarkets := []string{"Día", "Jumbo"}
	pharmacies := []string{"Farmacity", "FarmaOnline"}

	supermarketData := make(map[string]interface{})
	for _, store := range supermarkets {
		supermarketData[store] = getStoreStats(store)
	}

	pharmacyData := make(map[string]interface{})
	for _, store := range pharmacies {
		pharmacyData[store] = getStoreStats(store)
	}

	c.JSON(200, gin.H{
		"supermarkets": gin.H{
			"totalProducts": len(filterByStoreType(allProducts, true)),
			"stores":        supermarketData,
		},
		"pharmacies": gin.H{
			"totalProducts": len(filterByStoreType(allProducts, false)),
			"stores":        pharmacyData,
		},
	})
}

func CategoryStoresComparisonHandler(c *gin.Context) {
	mu.RLock()
	defer mu.RUnlock()

	result := make(map[string]interface{})
	stats := calculateCategoryStats()

	for category := range stats {
		products := filterByCategory(allProducts, category)

		supermarketStats := make(map[string]float64)
		pharmacyStats := make(map[string]float64)

		for _, p := range products {
			if isSupermarket(p.Store) {
				if _, exists := supermarketStats[p.Store]; !exists {
					supermarketStats[p.Store] = 0
				}
				supermarketStats[p.Store] += p.Price
			} else {
				if _, exists := pharmacyStats[p.Store]; !exists {
					pharmacyStats[p.Store] = 0
				}
				pharmacyStats[p.Store] += p.Price
			}
		}

		storeComparison := make(map[string]float64)
		for store, total := range supermarketStats {
			count := 0
			for _, p := range products {
				if p.Store == store {
					count++
				}
			}
			if count > 0 {
				storeComparison[store] = total / float64(count)
			}
		}
		for store, total := range pharmacyStats {
			count := 0
			for _, p := range products {
				if p.Store == store {
					count++
				}
			}
			if count > 0 {
				storeComparison[store] = total / float64(count)
			}
		}

		result[category] = storeComparison
	}

	c.JSON(200, result)
}

func CheapestByCategoryHandler(c *gin.Context) {
	category := c.Param("category")

	mu.RLock()
	defer mu.RUnlock()

	products := filterByCategory(allProducts, category)

	type ProductInfo struct {
		Name      string  `json:"name"`
		Price     float64 `json:"price"`
		Store     string  `json:"store"`
		Link      string  `json:"link"`
		Image     string  `json:"image"`
		Brand     string  `json:"brand"`
		ProductId string  `json:"productId"`
		Category  string  `json:"category"`
	}

	var productList []ProductInfo
	for _, p := range products {
		productList = append(productList, ProductInfo{
			Name:      p.Name,
			Price:     p.Price,
			Store:     p.Store,
			Link:      p.Link,
			Image:     p.Image,
			Brand:     p.Brand,
			ProductId: p.ProductId,
			Category:  p.Category,
		})
	}

	sort.Slice(productList, func(i, j int) bool {
		return productList[i].Price < productList[j].Price
	})

	limit := 10
	if len(productList) < limit {
		limit = len(productList)
	}

	c.JSON(200, gin.H{
		"category": category,
		"total":    len(productList),
		"cheapest": productList[:limit],
	})
}

File: prototype-next/lib/analytics.ts
import { Product, StoreStats, CategoryStats } from "@/types";
import { loadAllProducts } from "./data";

const SUPERMARKETS = ["Día", "Jumbo"];
const PHARMACIES = ["Farmacity", "FarmaOnline"];

export function isSupermarket(store: string): boolean {
  return SUPERMARKETS.includes(store);
}

export function filterByCategory(products: Product[], category: string): Product[] {
  return products.filter((p) => p.category === category);
}

export function filterByStore(products: Product[], store: string): Product[] {
  return products.filter((p) => p.store === store);
}

export function filterByStoreType(products: Product[], isSupermarketType: boolean): Product[] {
  return products.filter((p) => isSupermarket(p.store || "") === isSupermarketType);
}

export function getStoreBreakdown(products: Product[]): Record<string, StoreStats> {
  const storeStats: Record<string, StoreStats> = {};

  for (const p of products) {
    const stat = storeStats[p.store];
    if (stat) {
      stat.productCount++;
      stat.totalSpent += p.price;
      stat.minPrice = Math.min(stat.minPrice, p.price);
      stat.maxPrice = Math.max(stat.maxPrice, p.price);
    } else {
      storeStats[p.store] = {
        storeName: p.store,
        productCount: 1,
        averagePrice: p.price,
        minPrice: p.price,
        maxPrice: p.price,
        totalSpent: p.price,
      };
    }
  }

  for (const stat of Object.values(storeStats)) {
    stat.averagePrice = stat.totalSpent / stat.productCount;
  }

  return storeStats;
}

export function getPriceComparison(products: Product[]) {
  if (products.length === 0) {
    return {
      avgPrice: 0,
      minPrice: 0,
      maxPrice: 0,
      priceDiff: 0,
      cheapestStore: "",
    };
  }

  let totalPrice = 0;
  let minPrice = products[0].price;
  let maxPrice = products[0].price;
  let cheapestStore = products[0].store;

  for (const p of products) {
    totalPrice += p.price;
    if (p.price < minPrice) {
      minPrice = p.price;
      cheapestStore = p.store;
    }
    if (p.price > maxPrice) {
      maxPrice = p.price;
    }
  }

  return {
    avgPrice: totalPrice / products.length,
    minPrice,
    maxPrice,
    priceDiff: maxPrice - minPrice,
    cheapestStore,
  };
}

export function getStoreStats(products: Product[], storeName: string): StoreStats {
  const filtered = products.filter((p) => p.store === storeName);

  if (filtered.length === 0) {
    return {
      storeName,
      productCount: 0,
      averagePrice: 0,
      minPrice: 0,
      maxPrice: 0,
      totalSpent: 0,
    };
  }

  let totalSpent = 0;
  let minPrice = filtered[0].price;
  let maxPrice = filtered[0].price;

  for (const p of filtered) {
    totalSpent += p.price;
    minPrice = Math.min(minPrice, p.price);
    maxPrice = Math.max(maxPrice, p.price);
  }

  return {
    storeName,
    productCount: filtered.length,
    averagePrice: totalSpent / filtered.length,
    minPrice,
    maxPrice,
    totalSpent,
  };
}

export function calculateCategoryStats(products: Product[]): Record<string, CategoryStats> {
  const stats: Record<string, CategoryStats> = {};
  const categoryProducts: Record<string, Product[]> = {};

  for (const p of products) {
    const category = p.category || "Otros";
    if (!categoryProducts[category]) categoryProducts[category] = [];
    categoryProducts[category].push(p);
  }

  for (const [category, products] of Object.entries(categoryProducts)) {
    if (products.length === 0) continue;

    const stores = getStoreBreakdown(products);
    const priceComparison = getPriceComparison(products);

    stats[category] = {
      category,
      stores,
      cheapestStore: priceComparison.cheapestStore,
      averagePrice: priceComparison.avgPrice,
      minPrice: priceComparison.minPrice,
      maxPrice: priceComparison.maxPrice,
      priceDiff: priceComparison.priceDiff,
      productCount: products.length,
    };
  }

  return stats;
}

export async function getAllAnalytics() {
  const products = await loadAllProducts();
  const categories = calculateCategoryStats(products);

  return {
    products,
    categories,
    supermarkets: {
      totalProducts: filterByStoreType(products, true).length,
      stores: Object.fromEntries(
        SUPERMARKETS.map((store) => [store, getStoreStats(products, store)])
      ),
    },
    pharmacies: {
      totalProducts: filterByStoreType(products, false).length,
      stores: Object.fromEntries(
        PHARMACIES.map((store) => [store, getStoreStats(products, store)])
      ),
    },
  };
}
